util/pyclient/bft_client.py

```diff
@@ -173,58 +203,41 @@ async def send_to_primary(self, request):
             nursery.start_soon(self.sendto, request, (self.primary.ip,
                                                       self.primary.port))
 
-    async def send_all(self, request):
+    async def send_to_replicas(self, request, replicas):
         """Send a serialized request to all replicas"""
         async with trio.open_nursery() as nursery:
-            for replica in self.replicas:
+            for replica in replicas:
                 nursery.start_soon(self.sendto, request, (replica.ip,
                                                           replica.port))
+
     async def sendto(self, request, ip_port):
         """Send a request over a udp socket"""
         await self.sock.sendto(request, ip_port)
         self.msgs_sent += 1
 
-    async def recv(self, cancel_scope):
+    async def recv(self, required_quorum_size, cancel_scope):
         """
         Receive reply messages until a quorum is achieved or the enclosing
         cancel_scope times out.
         """
         while True:
             data, sender = await self.sock.recvfrom(self.config.max_msg_size)
-            header, reply = bft_msgs.unpack_reply(data)
+            sri_msg = rsi.MsgWithSpecificReplicaInfo(data, sender)
+            header, reply = sri_msg.get_common_reply()
             if self.valid_reply(header):
-                self.replies[sender] = (header, reply)
-            if self.has_quorum():
-                # This cancel will propagate upward and gracefully terminate all
-                # coroutines. self.reply will get set in self.has_quorum()
-                cancel_scope.cancel()
+                quorum_size = self.replies_manager.add_reply(sri_msg)
```

Comment:
I don't think this works exactly like it should. The problem is that the replies_manager has no notion of the specific MofN quorum. This allows replies to get added from replicas that are different from the ones required. My suggestion is to instead construct the replies_manager with the specific quorum per request so it can successfully add only replies from replicas in its required list.

Reply:
I prefer to separate the specific replica info infra from the m out of n quorums. I've added this check in the bft client itself such that a reply is valid only if came from one of the destination replicas